### DISCUSSION TITLE
Image::magick perl 5.35 macbook  m1 monterey 12.2.1

### DISCUSSION THREAD

**Discussion by kburdish** [ImageMagick/ImageMagick#4883] [Development]
Hey all!

I can't seem to get the above I cato work, i installed/uninstalled everything I can think of, 

Ive installed  imagemagick with homebrew,
cpanm Imsge::Magick fails with:
cc -c  -I/usr/local/include/ImageMagick-7 -DMAGICKCORE_HDRI_ENABLE=1 -DMAGICKCORE_QUANTUM_DEPTH=16 -I/usr/include/libxml2 -I"/usr/include/ImageMagick-7" -fno-common -DPERL_DARWIN -mmacosx-version-min=10.15 -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -DPERL_USE_SAFE_PUTENV -I/usr/include/freetype2 -g -O2 -Wall -pthread -DMAGICKCORE_HDRI_ENABLE=1 -DMAGICKCORE_QUANTUM_DEPTH=16 -Wno-error=implicit-function-declaration -O3   -DVERSION=\"7.0.11\" -DXS_VERSION=\"7.0.11\"  "-I/Users/kevin/perl5/perlbrew/perls/perl-5.35.1/lib/5.35.1/darwin-2level/CORE"  -D_LARGE_FILES=1 -DHAVE_CONFIG_H Magick.c
Magick.xs:56:10: fatal error: 'MagickCore/MagickCore.h' file not found
#include <MagickCore/MagickCore.h>
         ^~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
make: *** [Magick.o] Error 1

Any, and all help apprecaiated!!

**Comment by gwood** [comment]
### Installing `Image::Magick` on ARM silicon on macOS.

### Mixing Intel and ARM architecture causes issues in brew and cpanm builds

I realize this is a few years late to report this - but I wanted to document a successful strategy to the problem of ARM silicon failing to build Image::Magick when the proper `/opt/homebrew` path is attempted to be used, instead of the now incorrect `/usr/local` path such as when source has been upgraded from an Intel architecture in the past and left around to be found by makefiles that look in the standard places.

I had previously installed ImageMagick on both the `/usr/local` and `/opt/homebrew` paths as I had copied the home directory from an Intel host to an Arm host without wiping out the /usr/local with Intel libraries and executables.  [I started with an Arm simulator and slowly worked to replace the Intel with Arm.  This left a mess with respect to Arm and Intel collisions in this ImageMagick case.]

### Failure in build

```
LD_RUN_PATH="/usr/local/lib:/usr/lib" cc -Wl,-rpath,"/usr/local/lib" -Wl,-rpath,"/usr/lib" -L/usr/local/lib -lMagickCore-7.Q16HDRI -bundle -undefined dynamic_lookup -fstack-protector-strong   Magick.o  -o blib/arch/auto/Image/Magick/Magick.bundle  \
	   -L/usr/local/lib -lMagickCore-7.Q16HDRI -lm -L/opt/homebrew/opt/perl/lib/perl5/5.42/darwin-thread-multi-2level/CORE   \
	  
ld: warning: ignoring duplicate libraries: '-lMagickCore-7.Q16HDRI'
ld: warning: ignoring file '/usr/local/lib/libMagickCore-7.Q16HDRI.dylib': found architecture 'x86_64', required architecture 'arm64'

```
### Failure in test
```
$ make test
"/opt/homebrew/Cellar/perl/5.42.0/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- Magick.bs blib/arch/auto/Image/Magick/Magick.bs 644
PERL_DL_NONLAZY=1 "/opt/homebrew/Cellar/perl/5.42.0/bin/perl" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, 'blib/lib', 'blib/arch')" t/*.t
t/blob.t .......... Can't load '/Users/gww/.cpanm/work/1771176574.78958/Image-Magick-7.1.2/blib/arch/auto/Image/Magick/Magick.bundle' for module Image::Magick: dlopen(/Users/developer/.cpanm/work/1771176574.78958/Image-Magick-7.1.2/blib/arch/auto/Image/Magick/Magick.bundle, 0x0002): symbol not found in flat namespace '_AcquireAlignedMemory' at /opt/homebrew/opt/perl/lib/perl5/5.42/darwin-thread-multi-2level/DynaLoader.pm line 210.
� at t/blob.t line 20.
Compilation failed in require at t/blob.t line 20.
BEGIN failed--compilation aborted at t/blob.t line 20.
&Image::Magick::constant not defined. The required ImageMagick libraries are not installed or not installed properly.
END failed--call queue aborted at t/blob.t line 20.

```

The defaults in the `Image::Magick` `Makefile.PL` use `/usr/local `to do the compile instead of the correct `/opt/homebrew `path for Apple ARM Silicon.  The old Intel binaries conflict with the newer Arm binaries.

### Fixes to the environment:

I removed the older imagemagick installs from `/usr/local/lib` so they could not be found by the linker.

### Fixes to the Image::Magick build process

Then I changed each of the `/usr/local/` paths in the `Makefile.PL` so that they ALL pointed to `/opt/homebrew` instead.

This last change allowed the latest `Image::Magick` build and all of the tests to pass. These steps now all work on the Perl package that is downloaded by cpanm and mentioned in the log when the code fails.

```
perl Makefile.PL
make
make test
make install
```

To find the downloaded code to make the fixes below I looked at the error and found the log.

` ~/.cpanm/work/1771342686.32682/build.log`

Visited the directory
 
` ~/.cpanm/work/1771342686.32682`
to find the build directory.
`Image-Magick-7.1.2/`

I made changes - documented in the diff below.

Then built the Makefile, the code, tested and installed as described above.

diff -u Image-Magick-7.1.2{,-WORKING}/Makefile.PL

### The patch to Makefile.PL to correct these errors:

```
--- Image-Magick-7.1.2/Makefile.PL	2024-09-08 18:56:22
+++ Image-Magick-7.1.2-WORKING/Makefile.PL	2026-02-17 10:09:49
@@ -161,11 +161,11 @@
 }
 
 # defaults for LIBS & INC & CCFLAGS params that we later pass to Writemakefile
-my $INC_magick = '-I/usr/local/include/ImageMagick-7 -DMAGICKCORE_HDRI_ENABLE=1 -DMAGICKCORE_QUANTUM_DEPTH=16 -I/usr/include/libxml2 -I"' . $Config{'usrinc'} . '/ImageMagick-7"';
-my $LIBS_magick = '-L/usr/local/lib -lMagickCore-7.Q16HDRI -lm -L' . $Config{'archlib'} . '/CORE';
+my $INC_magick = '-I/opt/homebrew/include/ImageMagick-7 -DMAGICKCORE_HDRI_ENABLE=1 -DMAGICKCORE_QUANTUM_DEPTH=16 -I/usr/include/libxml2 -I"' . $Config{'usrinc'} . '/ImageMagick-7"';
+my $LIBS_magick = '-L/opt/homebrew/lib -lMagickCore-7.Q16HDRI -lm -L' . $Config{'archlib'} . '/CORE';
 my $CCFLAGS_magick = "$Config{'ccflags'} -I/usr/include/freetype2 -g -O2 -Wall -pthread -DMAGICKCORE_HDRI_ENABLE=1 -DMAGICKCORE_QUANTUM_DEPTH=16";
-my $LDFLAGS_magick   = "-L/usr/local/lib -lMagickCore-7.Q16HDRI $Config{'ldflags'} ";
-my $LDDLFLAGS_magick = "-L/usr/local/lib -lMagickCore-7.Q16HDRI $Config{'lddlflags'} ";
+my $LDFLAGS_magick   = "-L/opt/homebrew/lib -lMagickCore-7.Q16HDRI $Config{'ldflags'} ";
+my $LDDLFLAGS_magick = "-L/opt/homebrew/lib -lMagickCore-7.Q16HDRI $Config{'lddlflags'} ";
 
 if (($^O eq 'MSWin32') && ($Config{cc} =~ /gcc/)) {
   my($Ipaths, $Lpaths) = AutodetectWin32gcc();
@@ -239,7 +239,7 @@
 
    # Set Perl installation prefix to ImageMagick installation prefix
    # 'PREFIX'	=> '/usr/local',
-
+   PREFIX=> '/opt/homebrew',
    # Include delegate directories in tests
    test	=> { TESTS	=>	$delegate_tests},
 
@@ -263,6 +263,6 @@
      # tries to insert our library path first. Also, use the same compiler used
      # to build perlmain.c to link so that a C++ compiler may be used if
      # necessary.
-     $inherited =~ s:MAP_LINKCMD\s.*\s*\$\(CC\):MAP_LINKCMD   = \$(PERLMAINCC) -L/usr/local/lib: ;
+     $inherited =~ s:MAP_LINKCMD\s.*\s*\$\(CC\):MAP_LINKCMD   = \$(PERLMAINCC) -L/opt/homebrew/lib: ;
      $inherited;
  }
```

### Applying the patch

With the patch file in the directory above I applied that to the Makefile.PL like so:

` patch Makefile.PL < ../patch.patch`

**Reply by wollmers** [reply]
@gwood Thanks, works for me now.

My environment:
- MacBook Pro M5
- MacOS Tahoe 26.2
- homebrew
- perlbrew
- Perl (v5.43.10) built for darwin-2level

Essentially did the same changes in `Makefile.PL` before reading your HOWTO, but missed the two last changes in your patch.

Still get a warning:

```
~/github/perl/Image-Magick-7.1.2$ less /Users/helmut/.cpanm/work/1779815228.66155/build.log

ld: warning: -undefined dynamic_lookup may not work with chained fixups
ld: warning: dylib (/opt/homebrew/lib/libMagickCore-7.Q16HDRI.dylib) was built for newer macOS version (26.0) than being linked (13.1)
```

Test fails:

```
t/blob.t .......... ok
t/composite.t ..... ok
t/filter.t ........ 
Failed 1/58 subtests 
t/getattribute.t .. ok
t/montage.t ....... ok
t/ping.t .......... ok
t/read.t .......... ok
t/setattribute.t .. ok
t/write.t ......... ok

Test Summary Report
-------------------
t/filter.t      (Wstat: 0 Tests: 58 Failed: 1)
  Failed test:  2
Files=9, Tests=273,  1 wallclock secs ( 0.01 usr  0.01 sys +  0.35 cusr  0.05 csys =  0.42 CPU)
Result: FAIL
Failed 1/9 test programs. 1/273 subtests failed.
make: *** [test_dynamic] Error 255
```

Install with `--force`:

```
~/github/perl/Image-Magick-7.1.2$ cpanm --force .
--> Working on .
Configuring Image-Magick-v7.1.2 ... OK
Building and testing Image-Magick-v7.1.2 ... FAIL
! Testing Image-Magick-v7.1.2 failed but installing it anyway.
Successfully installed Image-Magick-v7.1.2
```

Examples in `./demo` work except 

```
display: delegate library support not built-in '' (X11) @ error/display.c/DisplayImageCommand/1907.
```

which I don't need in my codebase.
